File: Data-Structures-Algorithms-In-TS/Queue/QueueUsingStack.ts
class QueueUsingStack<T> {
    stk1: T[];
    stk2: T[];

    constructor() {
        this.stk1 = [];
        this.stk2 = [];
    }

    add(value: T): void {
        this.stk1.push(value);
    }

    length(): number {
        return this.stk1.length + this.stk2.length;
    }

    isEmpty(): boolean {
        return this.length() === 0;
    }

    remove(): T {
        if (this.stk2.length === 0) {
            while (this.stk1.length > 0) {
                const value = this.stk1.pop();
                this.stk2.push(value);
            }
        }
        return this.stk2.pop();
    }
}

// Testing code.
const que: QueueUsingStack<number> = new QueueUsingStack<number>();
que.add(1);
que.add(2);
que.add(3);
console.log(que.remove());
console.log(que.remove());

File: Data-Structures-and-Algorithms-using-Swift/Stack/QueueUsingStack.swift
class Stack<T> {
	var stk = [T]()
	
	public init() {}

	public var isEmpty : Bool {
		return stk.isEmpty
	}
	
	public var count : Int {
		return stk.count
	}

	public func push(_ value : T) {
		stk.append(value)
	}
	
	public func pop() -> T? {
		if stk.isEmpty  { 
			return nil 
		}
		return stk.removeLast()
	}
	
	public func top() -> T? {
		return stk.last
	}
}

class QueueUsingStack {
	var stk1 = Stack<Int>()
	var stk2 = Stack<Int>()

	public init() {}

	public func add(_ value : Int) {
		stk1.push(value)
	}

	public func remove() -> Int {
		if stk1.isEmpty == true && stk2.isEmpty == true {
			return 0
		}

		if stk2.isEmpty == true {
			while stk1.isEmpty == false {
				stk2.push(stk1.pop()!)
			}
		}
		return stk2.pop()!
	}
}

// Testing code
var que = QueueUsingStack()
que.add(1)
que.add(11)
que.add(111)
print(que.remove())
print(que.remove())
print(que.remove())
